src/sentry/migrations/0866_backfill_open_periods.py

```diff
@@ -0,0 +1,217 @@
+                )
+                end = None
+                end_activity = None
+    # Add the very first open period, which has no UNRESOLVED activity for the group creation
+    open_periods.append(
+        GroupOpenPeriod(
+            group_id=group_id,
+            project_id=project_id,
+            date_started=first_seen,
+            date_ended=end if end else None,
+            resolution_activity=end_activity,
+            user_id=end_activity.user_id if end_activity else None,
+        )
+    )
+    return open_periods
+
+
+def _backfill_group_open_periods(
+    apps: StateApps, group_data: list[tuple[int, datetime, int, int]]
+) -> None:
+    GroupOpenPeriod = apps.get_model("sentry", "GroupOpenPeriod")
+    Activity = apps.get_model("sentry", "Activity")
+
+    old_open_periods = list(
+        GroupOpenPeriod.objects.filter(group_id__in=[group_id for group_id, _, _, _ in group_data])
+        .values("group_id")
+        .annotate(Max("date_started"))
+    )
+    activities = list(
+        Activity.objects.filter(
+            group_id__in=[group_id for group_id, _, _, _ in group_data],
+            type__in=[ActivityType.SET_REGRESSION.value, ActivityType.SET_RESOLVED.value],
+        ).order_by("-datetime")
+    )
+
+    open_periods = []
+    for group_id, first_seen, status, project_id in group_data:
+        # Skip groups that already have open periods starting from the first seen date.
+        # These groups were either already backfilled or were created after the open period
+        # logic was added.
+        old_open_period = next(
+            (open_period for open_period in old_open_periods if open_period.group_id == group_id),
+            None,
+        )
+        if old_open_period and old_open_period.date_started == first_seen:
+            continue
+        query_end = old_open_period.date_started if old_open_period else None
+
+        activities_for_group = [
+            activity
+            for activity in activities
+            if activity.group_id == group_id
+            and (query_end is None or activity.datetime < query_end)
+        ]
+        # Backfill until the first open period that already exists.
+        open_periods.extend(
+            get_open_periods_for_group(
+                apps,
+                group_id,
+                status,
+                project_id,
+                first_seen,
+                activities_for_group,
+                GroupOpenPeriod,
+            )
+        )
+
+    GroupOpenPeriod.objects.bulk_create(open_periods)
+
+
+def backfill_group_open_periods(apps: StateApps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    Group = apps.get_model("sentry", "Group")
+
+    backfill_key = "backfill_group_open_periods_from_activity"
+    redis_client = redis.redis_clusters.get(settings.SENTRY_MONITORS_REDIS_CLUSTER)
+
+    progress_id = int(redis_client.get(backfill_key) or 0)
+    for group_data in chunked(
+        RangeQuerySetWrapperWithProgressBarApprox(
```

Comment:
Probably worth noting that the progress bar will be inaccurate whenever we're starting from a higher id.

If we ever cared to fix it, one way to kind of approximate better for this type of query would be to get the max id from the table, and just have the progress % be the distance of the current id from that id.

No need to fix unless you feel like it, jfyi